src/daemon/daemon.py

```diff
@@ -63,46 +66,115 @@ def _update_balance(db_utxo, addresses, new_utxo=True):
+        elif not reorg:
+            logging.warn("No backup to restore")
+            backup('""', '""', "/var/backups/mongodb/", mongo_backup_directory_path='/tmp/mongo_dump_backup', purge_local=30)
+            logging.info("Initial database state dumped at height: {}".format(get_highest_block_number_in_db()))
+        else:
+            logging.warn("No backup to restore and chain reorg occured. Need to fill DB from scratch.")
+            drop_db()
+    except Exception as e:
+        logging.warn(e)
+    
+    update_analyzer()
+
+def backup_db():
+    current_height = analyzer.max_block_number()
+    if current_height < BACKUP_FEATURE_HEIGHT:
+        return
+    
+    if current_height % 1000 == 0: # dump db every 1000th block
+        try:
+            logging.info("Doing backup at height: {}".format(current_height))
+            backup('""', '""', "/var/backups/mongodb/", mongo_backup_directory_path='/tmp/mongo_dump_backup', purge_local=30)
+        except Exception as e:
+            logging.warn(e)
+            
+def check_reorg():
+    current_height = analyzer.max_block_number()
+    if current_height < BACKUP_FEATURE_HEIGHT:
+        return False
     
+    reorg = not analyzer.check_chain()
+    if reorg:
+        restore_db(True)
+        
+    return reorg
+    
+def update_blockchain():
     def _reexecute_queries():
-        analyzer.set_blockchain(get_blockchain())
-        analyzer.set_addresses(get_addresses())
-        analyzer.set_utxos(get_utxos())
         analyzer.piechart_data(100, 1, analyzer.max_block_number())
         periods = ['week', 'month', 'all']
         for period in periods:
             interval, lower_height, upper_height = period_to_block_range_and_interval(period)
             analyzer.mining_difficulty(interval, lower_height, upper_height)
             analyzer.transaction_volume(interval, lower_height, upper_height)
     
-    start_block_number = _highest_block_number() + 1
+    start_block_number = analyzer.max_block_number() + 1
     end_block_number = get_block_count() + 1
     new_utxos = {}
     del_utxos = {}
     
-    for idx in range(start_block_number, end_block_number):
-        block = get_block(idx)
-        n_utxos = {}
-        d_utxos = {}
-        insert_block(block, n_utxos, d_utxos)
-        new_utxos = {**new_utxos, **n_utxos}
-        del_utxos = {**del_utxos, **d_utxos}
-    update_addresses(new_utxos, del_utxos)
+    try:
+        for idx in range(start_block_number, end_block_number):
+            logging.debug("Processing block = {}".format(idx))
+            
+            block = get_block(idx)
+            n_utxos = {}
+            d_utxos = {}
+            
+            insert_block(block, n_utxos, d_utxos)
+            new_utxos = {**new_utxos, **n_utxos}
+            del_utxos = {**del_utxos, **d_utxos}
+            update_analyzer()
+            
+            if check_reorg():
```

Comment:
It's not necessary to check for reorgs every block. I would do one check after processing the whole chunk, with a depth equal to `<size_of_the_chunk> + <some_constant>`

Reply:
Done.